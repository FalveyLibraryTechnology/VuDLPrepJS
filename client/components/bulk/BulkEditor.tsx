import React, { useEffect, useState } from "react";
import { baseUrl, objectDatastreamLicenseUrl } from "../../util/routes";
import { useFetchContext } from "../../context/FetchContext";
import { useEditorContext } from "../../context/EditorContext";
import BasicBreadcrumbs from "../shared/BasicBreadcrumbs";
import FormControl from "@mui/material/FormControl";
import InputLabel from "@mui/material/InputLabel";
import Select from "@mui/material/Select";
import MenuItem from "@mui/material/MenuItem";
import BlurSavingTextField from "../shared/BlurSavingTextField";

const BulkEditor = (): React.ReactElement => {
    const {
        state: { licensesCatalog },
        action: { initializeCatalog },
    } = useEditorContext();
    const [results, setResults] = useState("");
    const [selectedRecords, setSelectedRecords] = useState("");
    const [licenseKey, setLicenseKey] = useState("");
    const [query, setQuery] = useState("");
    const [selectedRecordIds, setSelectedRecordIds] = useState<Array<string>>([]);
    const {
        action: { fetchText },
    } = useFetchContext();

    useEffect(() => {
        initializeCatalog();
    }, []);

    const doFetchRecords = async () => {
        try {
            const result = await fetchText(
                baseUrl + "/api/edit/query/solr",
                { method: "POST", body: JSON.stringify({ query }) },
                { "Content-Type": "application/json" },
            );
            const json = JSON.parse(result);
            let text = "";
            const newSelectedRecordIds: Array<string> = [];
            if (json.numFound < 1) {
                text = "No results found.";
            } else {
                json.docs.forEach((doc) => {
                    text += `${doc.id}:\t${doc.title}\n`;
                    newSelectedRecordIds.push(doc.id);
                });
            }
            setSelectedRecords(text);
            setSelectedRecordIds(newSelectedRecordIds);
        } catch (error) {
            setSelectedRecords(error.message);
        }
    };

    const doApplyChanges = async () => {
        try {
            if (licenseKey == "") {
                setResults("No change requested.");
                return;
            }
            if (selectedRecordIds.length < 1) {
                setResults("No records selected.");
                return;
            }
            let result = "";
            for (let i = 0; i < selectedRecordIds.length; i++) {
                const id = selectedRecordIds[i];
                const text = await fetchText(
                    objectDatastreamLicenseUrl(id, "LICENSE"),
                    {
                        method: "POST",
                        body: JSON.stringify({
                            licenseKey,
                        }),
                    },
                    { "Content-Type": "application/json" },
                );
                result += `${id}: ${text}\n`;
                setResults(result);
            }
        } catch (error) {
            setResults(error.message);
        }
    };

    return (
        <div>
            <BasicBreadcrumbs />
            <h1>Bulk Editor</h1>
            <h2>Record Selector</h2>
            <FormControl fullWidth>
                <BlurSavingTextField
                    value={query}
                    setValue={setQuery}
                    options={{ id: "outlined-basic", label: "Search Query", variant: "outlined" }}
                />
            </FormControl>
            <FormControl>
                <button onClick={() => doFetchRecords()}>Fetch Records</button>
            </FormControl>
            <p>{`${selectedRecordIds.length} selected.`}</p>
            <pre id="selectedRecords">{selectedRecords}</pre>
            <h2>Changes to Apply</h2>
            <FormControl fullWidth>
                <InputLabel>Choose New License</InputLabel>
                <Select
                    label="Choose New License"
                    value={licenseKey}
                    onChange={(event) => setLicenseKey(event.target.value)}
                >
                    <MenuItem key="nochange" value="">
                        Do not change license.
                    </MenuItem>
                    {Object.entries(licensesCatalog).map(([key, license]) => {
                        return (
                            <MenuItem key={key} value={key}>
                                {license.name}
                            </MenuItem>
                        );
                    })}
                </Select>
            </FormControl>
            <FormControl>
                <button id="bulkEditSubmit" onClick={() => doApplyChanges()}>
                    Apply Changes
                </button>
            </FormControl>
            <h2>Results:</h2>
            <pre id="bulkEditResults">{results}</pre>
        </div>
    );
};

export default BulkEditor;
