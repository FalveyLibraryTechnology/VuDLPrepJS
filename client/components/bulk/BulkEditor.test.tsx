import React from "react";
import { beforeEach, describe, expect, it, jest } from "@jest/globals";
import { act } from "react-dom/test-utils";
import { fireEvent, render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import BulkEditor from "./BulkEditor";

const mockUseEditorContext = jest.fn();
jest.mock("../../context/EditorContext", () => ({
    useEditorContext: () => {
        return mockUseEditorContext();
    },
}));

const mockUseFetchContext = jest.fn();
jest.mock("../../context/FetchContext", () => ({
    useFetchContext: () => {
        return mockUseFetchContext();
    },
}));
jest.mock("../shared/BasicBreadcrumbs", () => () => "BasicBreadcrumbs");

describe("BulkEditor", () => {
    let editorValues;
    let fetchContextValues;
    beforeEach(() => {
        editorValues = {
            state: {
                licensesCatalog: {
                    testLicenseKey: {
                        name: "testLicense",
                    },
                },
            },
            action: {
                initializeCatalog: jest.fn(),
            },
        };
        fetchContextValues = {
            action: {
                fetchText: jest.fn(),
            },
        };
        mockUseEditorContext.mockReturnValue(editorValues);
        mockUseFetchContext.mockReturnValue(fetchContextValues);
    });

    it("renders", () => {
        const tree = renderer.create(<BulkEditor />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("reports failure if it receives bad JSON", async () => {
        render(<BulkEditor />);
        const input = screen.getByLabelText("Search Query");
        fireEvent.blur(input, {
            target: {
                value: "*:*",
            },
        });
        const fetchButton = screen.getByText("Fetch Records");
        fetchContextValues.action.fetchText.mockResolvedValueOnce("invalid JSON");
        await act(async () => {
            fireEvent.click(fetchButton);
        });
        expect(fetchContextValues.action.fetchText).toHaveBeenCalledWith(
            "http://localhost:9000/api/edit/query/solr",
            { body: '{"query":"*:*","rows":50}', method: "POST" },
            { "Content-Type": "application/json" },
        );
        const resultList = screen.getByTitle("Selected Records");
        expect(resultList.innerHTML).toEqual("Unexpected token i in JSON at position 0");
    });

    it("changes nothing if you submit without making selections", async () => {
        render(<BulkEditor />);
        const applyButton = screen.getByText("Apply Changes");
        await act(async () => {
            fireEvent.click(applyButton);
        });
        const resultList = screen.getByTitle("Bulk Edit Results");
        expect(resultList.innerHTML).toEqual("No change requested.");
    });

    it("reports empty Solr results and takes no action if they are submitted", async () => {
        render(<BulkEditor />);
        const input = screen.getByLabelText("Search Query");
        fireEvent.blur(input, {
            target: {
                value: "*:*",
            },
        });
        const fetchButton = screen.getByText("Fetch Records");
        fetchContextValues.action.fetchText.mockResolvedValueOnce('{"numFound": 0, "docs": []}');
        await act(async () => {
            fireEvent.click(fetchButton);
        });
        expect(fetchContextValues.action.fetchText).toHaveBeenCalledWith(
            "http://localhost:9000/api/edit/query/solr",
            { body: '{"query":"*:*","rows":50}', method: "POST" },
            { "Content-Type": "application/json" },
        );
        const recordList = screen.getByTitle("Selected Records");
        expect(recordList.innerHTML).toEqual("No results found.");
        const licenseControl = screen.getByRole("combobox");
        await act(async () => {
            fireEvent.mouseDown(licenseControl);
        });
        await act(async () => {
            fireEvent.click(screen.getByText("testLicense"));
        });
        const applyButton = screen.getByText("Apply Changes");
        await act(async () => {
            fireEvent.click(applyButton);
        });
        const resultList = screen.getByTitle("Bulk Edit Results");
        expect(resultList.innerHTML).toEqual("No records selected.");
    });

    it("performs a Solr search and changes a license", async () => {
        render(<BulkEditor />);
        const input = screen.getByLabelText("Search Query");
        fireEvent.blur(input, {
            target: {
                value: "*:*",
            },
        });
        const fetchButton = screen.getByText("Fetch Records");
        fetchContextValues.action.fetchText.mockResolvedValueOnce(
            '{"numFound": 2, "docs": [{"id": "foo", "title": "Foo"}, {"id": "bar", "title": "Bar"}]}',
        );
        await act(async () => {
            fireEvent.click(fetchButton);
        });
        expect(fetchContextValues.action.fetchText).toHaveBeenCalledWith(
            "http://localhost:9000/api/edit/query/solr",
            { body: '{"query":"*:*","rows":50}', method: "POST" },
            { "Content-Type": "application/json" },
        );
        const recordList = screen.getByTitle("Selected Records");
        expect(recordList.innerHTML).toEqual("foo:\tFoo\nbar:\tBar\n");

        const licenseControl = screen.getByRole("combobox");
        await act(async () => {
            fireEvent.mouseDown(licenseControl);
        });
        await act(async () => {
            fireEvent.click(screen.getByText("testLicense"));
        });
        const applyButton = screen.getByText("Apply Changes");
        fetchContextValues.action.fetchText.mockResolvedValueOnce("success");
        fetchContextValues.action.fetchText.mockResolvedValueOnce("failure");
        await act(async () => {
            fireEvent.click(applyButton);
        });
        expect(fetchContextValues.action.fetchText).toHaveBeenCalledWith(
            "http://localhost:9000/api/edit/object/foo/datastream/LICENSE/license",
            { body: '{"licenseKey":"testLicenseKey"}', method: "POST" },
            { "Content-Type": "application/json" },
        );
        expect(fetchContextValues.action.fetchText).toHaveBeenCalledWith(
            "http://localhost:9000/api/edit/object/bar/datastream/LICENSE/license",
            { body: '{"licenseKey":"testLicenseKey"}', method: "POST" },
            { "Content-Type": "application/json" },
        );
        const resultList = screen.getByTitle("Bulk Edit Results");
        expect(resultList.innerHTML).toEqual("(1/2) foo: success\n(2/2) bar: failure\n");
    });

    it("handles errors during license updates", async () => {
        render(<BulkEditor />);
        const input = screen.getByLabelText("Search Query");
        fireEvent.blur(input, {
            target: {
                value: "*:*",
            },
        });
        const fetchButton = screen.getByText("Fetch Records");
        fetchContextValues.action.fetchText.mockResolvedValueOnce(
            '{"numFound": 2, "docs": [{"id": "foo", "title": "Foo"}, {"id": "bar", "title": "Bar"}]}',
        );
        await act(async () => {
            fireEvent.click(fetchButton);
        });
        expect(fetchContextValues.action.fetchText).toHaveBeenCalledWith(
            "http://localhost:9000/api/edit/query/solr",
            { body: '{"query":"*:*","rows":50}', method: "POST" },
            { "Content-Type": "application/json" },
        );
        const recordList = screen.getByTitle("Selected Records");
        expect(recordList.innerHTML).toEqual("foo:\tFoo\nbar:\tBar\n");

        const licenseControl = screen.getByRole("combobox");
        await act(async () => {
            fireEvent.mouseDown(licenseControl);
        });
        await act(async () => {
            fireEvent.click(screen.getByText("testLicense"));
        });
        const applyButton = screen.getByText("Apply Changes");
        fetchContextValues.action.fetchText.mockResolvedValueOnce("success");
        fetchContextValues.action.fetchText.mockRejectedValue(new Error("kaboom"));
        await act(async () => {
            fireEvent.click(applyButton);
        });
        expect(fetchContextValues.action.fetchText).toHaveBeenCalledWith(
            "http://localhost:9000/api/edit/object/foo/datastream/LICENSE/license",
            { body: '{"licenseKey":"testLicenseKey"}', method: "POST" },
            { "Content-Type": "application/json" },
        );
        expect(fetchContextValues.action.fetchText).toHaveBeenCalledWith(
            "http://localhost:9000/api/edit/object/bar/datastream/LICENSE/license",
            { body: '{"licenseKey":"testLicenseKey"}', method: "POST" },
            { "Content-Type": "application/json" },
        );
        const resultList = screen.getByTitle("Bulk Edit Results");
        expect(resultList.innerHTML).toEqual("kaboom");
    });
});
