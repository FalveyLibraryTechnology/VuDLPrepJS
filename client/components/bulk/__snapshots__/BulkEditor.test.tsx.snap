// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BulkEditor renders 1`] = `
<div>
  BasicBreadcrumbs
  <h1>
    Bulk Editor
  </h1>
  <h2>
    Record Selector
  </h2>
  <div
    className="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
  >
    <div
      className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={false}
        htmlFor="outlined-basic"
        id="outlined-basic-label"
      >
        Search Query
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          disabled={false}
          id="outlined-basic"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-yjsfm1"
          >
            <span>
              Search Query
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
  <div
    className="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
  >
    <div
      className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={true}
        htmlFor="outlined-basic"
        id="outlined-basic-label"
      >
        Result Limit
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          disabled={false}
          id="outlined-basic"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value="50"
        />
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-14lo706"
          >
            <span>
              Result Limit
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
  <div
    className="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
  >
    <button
      onClick={[Function]}
    >
      Fetch Records
    </button>
  </div>
  <p>
    0 selected.
  </p>
  <pre
    id="selectedRecords"
    title="Selected Records"
  />
  <h2>
    Changes to Apply
  </h2>
  <div
    className="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={false}
    >
      Choose New License
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
      onClick={[Function]}
    >
      <div
        aria-controls=":r2:"
        aria-expanded="false"
        aria-haspopup="listbox"
        className="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="combobox"
        tabIndex={0}
      >
        <span
          className="notranslate"
        >
          â€‹
        </span>
      </div>
      <input
        aria-hidden={true}
        aria-invalid={false}
        className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        disabled={false}
        onAnimationStart={[Function]}
        onChange={[Function]}
        required={false}
        tabIndex={-1}
        value=""
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-yjsfm1"
        >
          <span>
            Choose New License
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    className="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
  >
    <button
      id="bulkEditSubmit"
      onClick={[Function]}
    >
      Apply Changes
    </button>
  </div>
  <h2>
    Results:
  </h2>
  <pre
    id="bulkEditResults"
    title="Bulk Edit Results"
  />
</div>
`;
